### PR TITLE
allow breaking before left parenthesis, not after

### PR DESCRIPTION
**Pull request by candux**
Some people and some organisations don't put a whitespace before a left parenthesis. Currently Maplibre only breaks after a left parenthesis, not before. 
This will put the left parenthesis at the end of the line.

You can notice this problem for example in "Freiburg(Breisgau) Hbf" which is the official station name.

Without this MR:
image

With this MR: 
image

## Launch Checklist

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [x] Include before/after visuals or gifs if this PR includes visual changes.
 - [x] Write tests for all new functionality.
 - [x] Add an entry to `CHANGELOG.md` under the `## main` section.

### CODE REVIEW

File: src/symbol/shaping.ts [hunk 1]
```diff
@@ -528,7 +534,7 @@ function determineLineBreaks(
         // surrounding spaces.
         if ((i < logicalInput.length() - 1)) {
             const ideographicBreak = charAllowsIdeographicBreaking(codePoint);
-            if (breakable[codePoint] || ideographicBreak || section.imageName) {
+            if (breakable[codePoint] || ideographicBreak || section.imageName || breakableBefore[logicalInput.getCharCode(i + 1)]) {
```
Comment (HarelM) [inline src/symbol/shaping.ts:537]: What will happen if this is the last character?

**Comment by wipfli** [comment]
Another special case that might be worth looking into are righ-to-left languages such as Arabic and Hebrew.

**Comment by HarelM** [comment]
True, there's an example page with RTL support, you can check how the following sentence will look:
היי (שלום) בעברית

File: src/symbol/shaping.ts (hunk 1 above)
Comment (candux) [reply src/symbol/shaping.ts]: thats a good point. I now check for that case and will not add a possible line break

**Comment by candux** [comment]
Thank you for your feedback. I checked some more cases and RTL languages

main:
image

with this MR:
image

I cannot find any regression after I fixed the case HarelM pointed out where a left parenthesis without a whitespace is at the end of the text.

I'm sure if the RTL-Plugin works correctly if RTL and LTR is mixed. The LTR Part is always at the beginning. Maybe somebody who knows more about it can create a bugreport for that.

```
<!DOCTYPE html>
<html>

<head>
  <meta charset="utf-8">
  <title>test</title>
  <meta name="viewport" content="initial-scale=1,maximum-scale=1,user-scalable=no">
  <script src="dist/maplibre-gl-dev.js"></script>
  <link href="dist/maplibre-gl.css" rel="stylesheet" />
  <style>
    body {
      margin: 0;
      padding: 0;
    }

    #map {
      position: absolute;
      top: 0;
      bottom: 0;
      width: 100%;
    }
  </style>
</head>

<body>
  <div id="map"></div>

<script>
    maplibregl.setRTLTextPlugin(
        'https://unpkg.com/@mapbox/mapbox-gl-rtl-text@0.2.3/mapbox-gl-rtl-text.min.js',
        true // Lazy load the plugin
    );
  const map = new maplibregl.Map({
    container: 'map',
    style: {
      version: 8,
      sources: {},
      layers: [],

      glyphs:"https://api.maptiler.com/fonts/{fontstack}/{range}.pbf?key=AKWk1YMvUCQsX9tcAiRg"
    },
    "center": [
      7.8511,
      47.9877
    ],
    "zoom": 14
  });


  texts = [ "oooooooooooo (baaaaaaar)", "oooooooooooo(baaaaaaar)", "ooooooooooooaaaaaaa(", "ooooooooooooaaaaaaa (", "oooooooooaaaaaaa ()", "ב(שלוםשלוםשלוםשלום)עבריתבעברית", "ב (שלוםשלוםשלוםשלום) עבריתבעברית", "(שלוםשלוםשלוםשלום) בעבריתבעברית", "(שלוםשלוםשלוםשלום)בעבריתבעברית", "Point 10"]

  map.on('load', function () {
    const geojson = {
      "type": "FeatureCollection",
      "features": Array.from({ length: 10 }, (_, i) => ({
        "type": "Feature",
        "geometry": {
          "type": "Point",
          "coordinates": [7.8411 + i * 0.004, 47.9977 - i * 0.002]
        },
        "properties": {
          "text": texts[i]
        }
      }))
    };

    // Add the source to the map
    map.addSource('points', {
      "type": "geojson",
      "data": geojson
    });

    // Add a layer to display the text attribute of each point
    map.addLayer({
      "id": "points-text",
      "type": "symbol",
      "source": "points",
      "layout": {
        "text-field": ["get", "text"],
        "text-size": 14
      }
    });
  });
</script>

</body>

</html>

```

**Comment by wipfli** [comment]
Does the behavior for labels like `Basel-Stadt` change?

**Comment by candux** [comment]
> Does the behavior for labels like `Basel-Stadt` change?

no. I tested it with many Labels on real maps. Only the intended behavior on left parenthesis changes.

File: CHANGELOG.md [hunk 2]
```diff
@@ -10,11 +10,10 @@
 - _...Add new stuff here..._
 
 ### 🐞 Bug fixes
-
+- allow breaking lines in labels before a left parenthesis
```
Comment (HarelM) [inline CHANGELOG.md:13]:
```suggestion
- Allow breaking lines in labels before a left parenthesis ([#4138](https://github.com/maplibre/maplibre-gl-js/pull/4138))
```

**Comment by HarelM** [comment]
Overall, this looks good.
I need to find a way to add the coverage report to these tests as well as right now it seems like the coverage is missing, although I don't think this is the case, as it looks like the test is covering the code you added.
Let me try and solve this before this is merged.

File: test/integration/symbol-shaping/shaping.test.ts [hunk 3]
```diff
@@ -87,6 +87,15 @@ describe('shaping', () => {
     if (UPDATE) fs.writeFileSync(path.resolve(__dirname, './tests/text-shaping-zero-width-space.json'), JSON.stringify(shaped, null, 2));
     expect(shaped).toEqual(expectedZeroWidthSpaceBreak);
 
+    // break before left parenthesis without whitespace
+    // input: oooooooo(aaaaaaaaa)
+    // output: oooooooo\n(aaaaaaaaa)
+    const expectedBreakBeforeBracket = require('./tests/text-shaping-break-before-bracket.json');
+
+    shaped = shapeText(Formatted.fromString('oooooooo(aaaaaaaaa)'), glyphs, glyphPositions, images, fontStack, 5 * oneEm, oneEm, 'center', 'center', 0, [0, 0], WritingMode.horizontal, false, 'point', layoutTextSize, layoutTextSizeThisZoom);
```
Comment (HarelM) [inline test/integration/symbol-shaping/shaping.test.ts:95]:
```suggestion
    shaped = shapeText(Formatted.fromString('oooooooo(aaaaaaaaa)'), glyphs, glyphPositions, images, fontStack, 5 * oneEm, oneEm, 'center', 'center', 0, [0, 0], WritingMode.horizontal, false, layoutTextSize, layoutTextSizeThisZoom);
```

**Comment by HarelM** [comment]
I've created the following PR:
- #4171

Once merged to main and then merged here, we should be able to see the coverage of the integration tests around the shaping part.

**Comment by HarelM** [comment]
I've merged the above PR to main and now merged main to this branch.
Let's see what the code coverage returns.

File: CHANGELOG.md [hunk 4]
```diff
@@ -10,11 +10,10 @@
 - _...Add new stuff here..._
 
 ### 🐞 Bug fixes
-
+- Allow breaking lines in labels before a left parenthesis ([#4138](https://github.com/maplibre/maplibre-gl-js/pull/4138))
```
Comment (HarelM) [inline CHANGELOG.md:13]:
```suggestion
- ⚠️ Allow breaking lines in labels before a left parenthesis ([#4138](https://github.com/maplibre/maplibre-gl-js/pull/4138))
```

**Comment by wipfli** [comment]
This change looks good to me as well. Thanks @candux.

Again, similar to the now allowed line breaking on text along lines, this pull request is a breaking change of how the map looks like. I think @HarelM indicates this with the "!" in the changelog... In any case, @louwers / @HarelM please coordinate that there is a corresponding issue in MapLibre Native for breaking visual changes in MapLibre GL JS. Thank you!

**Comment by louwers** [comment]
Can a render test be added? I'm tracking disparity issues based on render tests.

**Comment by HarelM** [comment]
We've added shaping tests, but a render test with the above style should be easy to add. 
@candux any chance you can asd a render test with the scenario you highlighted? It's very thorough I think.

**Comment by candux** [comment]
@HarelM @louwers sure. I pushed two render tests. text-breaking/left-parenthesis highlights the change

**Comment by HarelM** [comment]
Great, thanks!

**Review by HarelM — APPROVED** [review]

**Comment by louwers** [comment]
Thank you!

I updated the [Render Test Parity Status Report](https://gist.github.com/louwers/0259c83872da7093670627a43f32b4a1) with a tracking issue.

**Comment by louwers** [comment]
@candux Bahn.de lists the station name with space by the way https://www.bahnhof.de/freiburg-breisgau-hbf

In Dutch, you would not be allowed to break off a word that contains parenthesis.

```
Freiburg(Breisgau)
Hbf
```

would be more correct, but I'm not sure if that fits?

Native already penalizes open parenthesis at the end of the line, I'll try penalizing it more and see what happens.

```cpp
    // Penalize open parenthesis at end of line
    if (codePoint == 0x28 || codePoint == 0xff08) {
        penalty += 50;
    }
```

**Comment by candux** [comment]
@louwers 
> Bahn.de lists the station name with space by the way https://www.bahnhof.de/freiburg-breisgau-hbf

We use the timetables for the official names. The timetables from Delfi and SBB unfortunately don't have a space.

> In Dutch, you would not be allowed to break off a word that contains parenthesis.

Oh, I didn't know that.

> Native already penalizes open parenthesis at the end of the line, I'll try penalizing it more and see what happens.

I don't think that is necessary. Just don't put "(" in the list of characters where a break after the character is allowed. I just found this unicode document which also says the same: https://www.unicode.org/reports/tr14/

If you want, I can create a PR to not break before the "(". Then it would be
```
Freiburg(Breisgau)
Hbf
```

**Pull request merged**